lib/preprocessors/glimmer.js

```diff
@@ -101,49 +117,113 @@ function mapRange(messages, filename) {
-      return message;
-    }
+    // 1. handle eslint diagnostics on single lines.
+    if (message.line === message.endLine) {
+      const originalLine = originalLines[message.line - 1];
+      const transformedLine = lines[message.line - 1];
+      if (originalLine === transformedLine) {
+        // original line and transformed line match exactly. return original diagnostic message
+        return message;
+      }
 
-    const line = lines[message.line - 1];
-    const token = line.slice(message.column - 1, message.endColumn - 1);
-
-    // Now that we have the token, we need to find the location
-    // in the original text
-    //
-    // TODO: Long term, we should use glimmer syntax parsing to find
-    // the correct node -- otherwise it's too easy to
-    // partially match on similar text
-    let originalLineNumber = 0;
-    let originalColumnNumber = 0;
-
-    for (const [index, line] of originalLines.entries()) {
-      const column = line.search(new RegExp(`\\b${escapeRegExp(token)}\\b`));
-      if (column > -1) {
-        originalLineNumber = index + 1;
-        originalColumnNumber = column + 1;
-        break;
+      // when the lines do not match, we've hit a lint error on the line containing the
+      // <template> tag, meaning we need to re-work the message
+      const modified = { ...message };
+      let token = transformedLine.slice(message.column - 1, message.endColumn - 1);
+
+      // lint error is specifically on JUST the opening <template> tag
+      if (openingTemplateTagRegex.test(token)) {
+        token = `<${util.TEMPLATE_TAG_NAME}>`;
+
+        // this case is simple: we know the starting column will be correct, so we just
+        // need to adjust the endColumn for the difference in length between the transformed
+        // token and the original token.
+        modified.endColumn = modified.column + token.length;
+      } else if (oneLineTemplateRegex.test(token)) {
+        // lint error is on a full, one-line <template>foo</template>
+        const templateContext = token.match(oneLineTemplateRegex)[1];
+        token = `<${util.TEMPLATE_TAG_NAME}>${templateContext}<${util.TEMPLATE_TAG_NAME}>`;
+
+        // this case is simple: we know we have a one-line template invocation, and the
+        // start `column` will be the same regardless of syntax. simply calculate the
+        // length of the full token `<template>...</template>` and set the endColumn.
+        modified.endColumn = modified.column + token.length;
+      } else {
+        // if we haven't fallen into one of the cases above, we are likely dealing with
+        // a scope token in the template placeholder, typically for being used but not
+        // defined. the `token` should be the specific template API which the error is on,
+        // so we need to find the usage of the token in the template
+
+        // TODO: this is still bug-prone, and should be refactored to do something like:
```

Comment:
I think we kinda have two options to making this less bug prone:
 - re-work template-imports / glimmer-parse / getTemplateLocals to get us positions 
  - upstream `<template>` to TC39, making our lives easier across the whole ecosystem. https://wycats.github.io/polaris-sketchwork/content-tag/second-phase-compiler.html